ghcide/src/Development/IDE/LSP/LanguageServer.hs

```diff
@@ -176,28 +177,32 @@ setupLSP  recorder getHieDbLoc userHandlers getIdeState clientMsgVar = do
         -- Cancel requests are special since they need to be handled
         -- out of order to be useful. Existing handlers are run afterwards.
 
-  let doInitialize = handleInit recorder getHieDbLoc getIdeState reactorLifetime exit clearReqId waitForCancel clientMsgChan
+  let doInitialize = handleInit root recorder getHieDbLoc getIdeState reactorLifetime exit clearReqId waitForCancel clientMsgChan
 
   let interpretHandler (env,  st) = LSP.Iso (LSP.runLspT env . flip (runReaderT . unServerM) (clientMsgChan,st)) liftIO
 
   pure (doInitialize, asyncHandlers, interpretHandler)
 
 
 handleInit
-    :: Recorder (WithPriority Log)
+    :: FilePath
+    -> Recorder (WithPriority Log)
     -> (FilePath -> IO FilePath)
-    -> (LSP.LanguageContextEnv config -> Maybe FilePath -> WithHieDb -> IndexQueue -> IO IdeState)
+    -> (LSP.LanguageContextEnv config -> FilePath -> WithHieDb -> IndexQueue -> IO IdeState)
     -> MVar ()
     -> IO ()
     -> (SomeLspId -> IO ())
     -> (SomeLspId -> IO ())
     -> Chan ReactorMessage
     -> LSP.LanguageContextEnv config -> TRequestMessage Method_Initialize -> IO (Either err (LSP.LanguageContextEnv config, IdeState))
-handleInit recorder getHieDbLoc getIdeState lifetime exitClientMsg clearReqId waitForCancel clientMsgChan env (TRequestMessage _ _ m params) = otTracedHandler "Initialize" (show m) $ \sp -> do
+handleInit rootDir recorder getHieDbLoc getIdeState lifetime exitClientMsg clearReqId waitForCancel clientMsgChan env (TRequestMessage _ _ m params) = otTracedHandler "Initialize" (show m) $ \sp -> do
     traceWithSpan sp params
-    let root = LSP.resRootPath env
-    dir <- maybe getCurrentDirectory return root
-    dbLoc <- getHieDbLoc dir
+    let rootMaybe = LSP.resRootPath env
+    -- only shift if lsp root is different from the rootDir
+    when (rootMaybe /= Just rootDir) $ do
```

Comment:
I'm a bit confused about what's going on here. Why do we have two sources of root directory? Can't we just use the one from the LSP setup?

Reply:
Lsp have a maybe one specified by the client, wrapper and exe could pass in one as the default one.

Reply:
Okay, so the idea is:
- The wrapper or top-level exe will do `getCurrentDirectory` once, and then pass that around as the root
- If we get a root directory from the client, then we will cd into that directory
- After that, we need to keep around whatever root directory we are using so we can use it to make paths absolute

Is that right? I wonder if we should implement this on the `lsp` side. We could then make it so that the `LspEnv` always contains a root directory, either passed in or taken from the client.

Reply:
That's right. 
For `lsp` side implmentation.
I suppose that would not make much different even if we pass the one from server to initilized the `LspEnv`. Most of places we use root are in `sessionLoader` and `Rules` system. So the only time we need to query the `LspEnv`'s root is here, when we are doing initialization for shake.

Reply:
I think maybe this is already not correct. `rootUri` is deprecated in favour of `workspaceFolders`: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_workspaceFolders

So in general the server might have several open folders. Which means there is no single answer to "how do I make a path absolute". I hope that means that clients won't send us many relative paths... Perhaps we should instead just stick with our original working directory?

Reply:
I see,In fact, I would very much like to remove the make a path absolute thing,
The initial goal here is to remove the dependency on `cwd` for our test as much as possible.
So it's more like a temperal thing if we wish to still support `rootUri` and remove the `cwd` lock.

Reply:
Yes, I'm not really sure what we're supposed to do with the workspace folders. I think maybe the thing to do for now would just be to preserve the "original working directory" and thread it around so we don't need to change directory as you say. So maybe just drop this bit?

Reply:
Do you mean we drop the `setCurrentDirectory` entirly here?
That's I originally did, but I found we still have some tests that is dependent on shift the cwd to lsp's rootdir. 

The original `setCurrentDirectory` is in `getIdeState`. We used to do the shift no matter the rootdir is comming from defaultmain's arguement or lsp' root. It create a problem for threadings with the shift.

So I move it up here instead of completely dropping it,  it means we only shift if lsp's root and defaultmain's argument's root have conflict. And hence remove the most shift cases in the test.